{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `paper_metadata.csv` file was downloaded from [here](https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-019-10092-5/MediaObjects/41467_2019_10092_MOESM5_ESM.xlsx)  \n",
    "It was converted to csv, the first two lines were removed and the values in Sequencing ID where used to fill the ENA accession column where it was missing  \n",
    "The values in ENA accession of the Ashton Study were used to download the sequencing data with the downloading-tools workflow.  \n",
    "The `reads_tables.csv` file was created by the downloading-tools workflow.  \n",
    "The run ERR2624135 was not possible to download.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "original = pd.read_csv('paper_metadata.csv', header = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads = pd.read_csv('reads_table.csv', header = 0)\n",
    "reads = reads[['sample', 'run']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the metadata and the read accessions to have a sample name and get the strain name from the Sequencing ID column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = pd.merge(original, reads, left_on = 'ENA accession', right_on = 'run', how = 'left')\n",
    "joined = joined.drop(columns = ['run'])\n",
    "joined = joined.rename(columns = {'Sequencing ID': 'strain', 'Isolation source': 'source', 'Sub-clade' : 'VNI_subdivision', 'Country of origin': 'Country_of_origin', 'ENA accession': 'SRA_Accession'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign lineage from the colummn `Species ID from mash analysis`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined.loc[joined['Species ID from mash anlaysis'] == 'Cryptococcus neoformans var. grubii H99', 'lineage'] = 'VNI'\n",
    "joined.loc[joined['Species ID from mash anlaysis'] == 'Cryptococcus neoformans var. grubii H99/Cryptococcus neoformans var. neoformans JEC21 hybrid', 'lineage'] = 'AD_hybrid'\n",
    "joined.loc[joined['Species ID from mash anlaysis'] == 'Cryptococcus gattii WM276', 'lineage'] = 'gattii'\n",
    "joined.loc[joined['VNI_subdivision'] == 'VNII', 'lineage'] = 'VNII'\n",
    "joined.loc[joined['VNI_subdivision'] == 'VNII', 'VNI_subdivision'] = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Put the columns `sample`, `strain`, `lineage` and `VNI_subdivision` to the left of the rest and sort by VNI_subdivision"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = joined[['sample', 'strain','lineage','source', 'VNI_subdivision'] + [col for col in joined.columns if col not in ['sample', 'strain','lineage','source','VNI_subdivision'] ]]\n",
    "joined = joined.sort_values('VNI_subdivision')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Print number of samples in each lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>lineage</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>VNI</td>\n",
       "      <td>678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>VNII</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>gattii</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Desjardins</td>\n",
       "      <td>VNI</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Study    lineage  counts\n",
       "0      Ashton  AD_hybrid       5\n",
       "1      Ashton        VNI     678\n",
       "2      Ashton       VNII       4\n",
       "3      Ashton     gattii      12\n",
       "4  Desjardins        VNI     185"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for col in joined.select_dtypes(include='object').columns:\n",
    "    joined[col] = joined[col].astype('category')\n",
    "joined.groupby(['Study', 'lineage'], observed=True).size().reset_index(name='counts')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make file with metadata of all samples in Ashton paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined.to_csv('metadata_all_ashton_and_vni_desj.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out the Desjardins samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ashton = joined[joined['Study'] != 'Desjardins']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out the missing sample that was not downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "ashton = ashton[ashton['SRA_Accession'] != 'ERR2624135']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>strain</th>\n",
       "      <th>lineage</th>\n",
       "      <th>source</th>\n",
       "      <th>VNI_subdivision</th>\n",
       "      <th>Lab ID</th>\n",
       "      <th>Species ID from mash anlaysis</th>\n",
       "      <th>Study</th>\n",
       "      <th>HIV status</th>\n",
       "      <th>SRA_Accession</th>\n",
       "      <th>Country_of_origin</th>\n",
       "      <th>Continent of Origin</th>\n",
       "      <th>Year of Origin</th>\n",
       "      <th>Mean depth of mapping with MQ &gt; 30 across whole genome</th>\n",
       "      <th>Proportion of genome covered by at least 5 reads which mapped with MQ &gt; 30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>ERS542448</td>\n",
       "      <td>14936_1#57</td>\n",
       "      <td>VNI</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>VNIa-32</td>\n",
       "      <td>BK167</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR842672</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>31.391425</td>\n",
       "      <td>0.959204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>ERS2540945</td>\n",
       "      <td>04CN-65-072</td>\n",
       "      <td>VNI</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>VNIa-32</td>\n",
       "      <td>04CN-65-072</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624432</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>54.041513</td>\n",
       "      <td>0.990784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>ERS2541049</td>\n",
       "      <td>04CN-65-161</td>\n",
       "      <td>VNI</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>VNIa-32</td>\n",
       "      <td>04CN-65-161</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624145</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.413277</td>\n",
       "      <td>0.991486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>ERS2541305</td>\n",
       "      <td>04CN-63-036</td>\n",
       "      <td>VNI</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>VNIa-32</td>\n",
       "      <td>04CN-63-036</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624319</td>\n",
       "      <td>Malawi</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>49.200974</td>\n",
       "      <td>0.987908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>ERS1142749</td>\n",
       "      <td>20427_2#52</td>\n",
       "      <td>VNI</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>VNIa-32</td>\n",
       "      <td>BK92</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR1671650</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>38.213466</td>\n",
       "      <td>0.960260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>ERS2541337</td>\n",
       "      <td>BMD_1385</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_1385</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624309</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>27.804298</td>\n",
       "      <td>0.608818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>ERS2541338</td>\n",
       "      <td>BMD_1516</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_1516</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624397</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>25.317098</td>\n",
       "      <td>0.601971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>ERS2541339</td>\n",
       "      <td>BMD_2014</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_2014</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624500</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.731607</td>\n",
       "      <td>0.587846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>ERS2541082</td>\n",
       "      <td>BMD1964</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD1964</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624244</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>35.464555</td>\n",
       "      <td>0.601434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>ERS2541145</td>\n",
       "      <td>BMD2233</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD2233</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624259</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>27.246364</td>\n",
       "      <td>0.600599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>698 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample       strain lineage    source VNI_subdivision       Lab ID  \\\n",
       "342   ERS542448   14936_1#57     VNI  Clinical         VNIa-32        BK167   \n",
       "245  ERS2540945  04CN-65-072     VNI  Clinical         VNIa-32  04CN-65-072   \n",
       "275  ERS2541049  04CN-65-161     VNI  Clinical         VNIa-32  04CN-65-161   \n",
       "150  ERS2541305  04CN-63-036     VNI  Clinical         VNIa-32  04CN-63-036   \n",
       "547  ERS1142749   20427_2#52     VNI  Clinical         VNIa-32         BK92   \n",
       "..          ...          ...     ...       ...             ...          ...   \n",
       "555  ERS2541337     BMD_1385  gattii  Clinical             NaN     BMD_1385   \n",
       "556  ERS2541338     BMD_1516  gattii  Clinical             NaN     BMD_1516   \n",
       "557  ERS2541339     BMD_2014  gattii  Clinical             NaN     BMD_2014   \n",
       "581  ERS2541082      BMD1964  gattii  Clinical             NaN      BMD1964   \n",
       "586  ERS2541145      BMD2233  gattii  Clinical             NaN      BMD2233   \n",
       "\n",
       "               Species ID from mash anlaysis   Study HIV status SRA_Accession  \\\n",
       "342  Cryptococcus neoformans var. grubii H99  Ashton   positive     ERR842672   \n",
       "245  Cryptococcus neoformans var. grubii H99  Ashton   positive    ERR2624432   \n",
       "275  Cryptococcus neoformans var. grubii H99  Ashton   positive    ERR2624145   \n",
       "150  Cryptococcus neoformans var. grubii H99  Ashton   positive    ERR2624319   \n",
       "547  Cryptococcus neoformans var. grubii H99  Ashton   positive    ERR1671650   \n",
       "..                                       ...     ...        ...           ...   \n",
       "555                Cryptococcus gattii WM276  Ashton   negative    ERR2624309   \n",
       "556                Cryptococcus gattii WM276  Ashton   negative    ERR2624397   \n",
       "557                Cryptococcus gattii WM276  Ashton   negative    ERR2624500   \n",
       "581                Cryptococcus gattii WM276  Ashton   negative    ERR2624244   \n",
       "586                Cryptococcus gattii WM276  Ashton   negative    ERR2624259   \n",
       "\n",
       "    Country_of_origin Continent of Origin  Year of Origin  \\\n",
       "342           Vietnam                Asia          2006.0   \n",
       "245            Uganda              Africa          2013.0   \n",
       "275            Uganda              Africa             NaN   \n",
       "150            Malawi              Africa          2014.0   \n",
       "547           Vietnam                Asia          2005.0   \n",
       "..                ...                 ...             ...   \n",
       "555           Vietnam                Asia          2007.0   \n",
       "556           Vietnam                Asia          2008.0   \n",
       "557           Vietnam                Asia             NaN   \n",
       "581           Vietnam                Asia          2009.0   \n",
       "586           Vietnam                Asia          2010.0   \n",
       "\n",
       "     Mean depth of mapping with MQ > 30 across whole genome  \\\n",
       "342                                          31.391425        \n",
       "245                                          54.041513        \n",
       "275                                          45.413277        \n",
       "150                                          49.200974        \n",
       "547                                          38.213466        \n",
       "..                                                 ...        \n",
       "555                                          27.804298        \n",
       "556                                          25.317098        \n",
       "557                                          20.731607        \n",
       "581                                          35.464555        \n",
       "586                                          27.246364        \n",
       "\n",
       "     Proportion of genome covered by at least 5 reads which mapped with MQ > 30  \n",
       "342                                           0.959204                           \n",
       "245                                           0.990784                           \n",
       "275                                           0.991486                           \n",
       "150                                           0.987908                           \n",
       "547                                           0.960260                           \n",
       "..                                                 ...                           \n",
       "555                                           0.608818                           \n",
       "556                                           0.601971                           \n",
       "557                                           0.587846                           \n",
       "581                                           0.601434                           \n",
       "586                                           0.600599                           \n",
       "\n",
       "[698 rows x 15 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ashton"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>lineage</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>VNI</td>\n",
       "      <td>677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>VNII</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ashton</td>\n",
       "      <td>gattii</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Study    lineage  counts\n",
       "0  Ashton  AD_hybrid       5\n",
       "1  Ashton        VNI     677\n",
       "2  Ashton       VNII       4\n",
       "3  Ashton     gattii      12"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ashton.groupby(['Study', 'lineage'], observed=True).size().reset_index(name='counts')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "ashton.to_csv('metadata.csv', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make file with non-VNI samples to exclude from the FungalPop analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ashton[ashton['lineage'] != 'VNI']\n",
    "exclude = exclude.sort_values('lineage')\n",
    "exclude.to_csv('metadata_ashton_non_vni.csv', index = False, header = True)\n",
    "exclude['sample'].to_csv('exclude.txt', index = False, header = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare names in the Ashton phylogeny to metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "865\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GCF_000149245', '04CN-63-018']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree = Phylo.read('/FastData/czirion/Crypto_Diversity_Pipeline/analyses/data/raw/2017.06.09.all_ours_and_desj.snp_sites.mod.fa.cln.tree', 'newick')\n",
    "tips =[tip.name for tip in tree.get_terminals()]\n",
    "print(len(tips))\n",
    "tips_not_in_metadata = [tip for tip in tips if tip not in joined['strain'].values]\n",
    "tips_not_in_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>strain</th>\n",
       "      <th>lineage</th>\n",
       "      <th>source</th>\n",
       "      <th>VNI_subdivision</th>\n",
       "      <th>Lab ID</th>\n",
       "      <th>Species ID from mash anlaysis</th>\n",
       "      <th>Study</th>\n",
       "      <th>HIV status</th>\n",
       "      <th>SRA_Accession</th>\n",
       "      <th>Country_of_origin</th>\n",
       "      <th>Continent of Origin</th>\n",
       "      <th>Year of Origin</th>\n",
       "      <th>Mean depth of mapping with MQ &gt; 30 across whole genome</th>\n",
       "      <th>Proportion of genome covered by at least 5 reads which mapped with MQ &gt; 30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERS2541256</td>\n",
       "      <td>04CN-03-031</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-03-031</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99/Crypto...</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624093</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>45.693733</td>\n",
       "      <td>0.975470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERS2541170</td>\n",
       "      <td>04CN-03-074</td>\n",
       "      <td>VNII</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-03-074</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624467</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>43.472265</td>\n",
       "      <td>0.965882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERS2541130</td>\n",
       "      <td>04CN-03-088</td>\n",
       "      <td>VNII</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-03-088</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624180</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>45.057615</td>\n",
       "      <td>0.966880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERS2540936</td>\n",
       "      <td>04CN-63-006</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-63-006</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624156</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>30.811483</td>\n",
       "      <td>0.611245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERS2540975</td>\n",
       "      <td>04CN-63-020</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-63-020</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624413</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>31.976266</td>\n",
       "      <td>0.604924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERS2541070</td>\n",
       "      <td>04CN-63-021</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-63-021</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624263</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>30.489235</td>\n",
       "      <td>0.602266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERS2541050</td>\n",
       "      <td>04CN-64-092</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-64-092</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99/Crypto...</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624285</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>32.871236</td>\n",
       "      <td>0.969852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERS2541321</td>\n",
       "      <td>04CN-65-019</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-019</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99/Crypto...</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624434</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>78.893862</td>\n",
       "      <td>0.978027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ERS2541042</td>\n",
       "      <td>04CN-65-031</td>\n",
       "      <td>VNII</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-031</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624151</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>31.248864</td>\n",
       "      <td>0.963734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ERS2541325</td>\n",
       "      <td>04CN-65-051</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-051</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624462</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>44.914011</td>\n",
       "      <td>0.616338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERS2540953</td>\n",
       "      <td>04CN-65-104</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-104</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624451</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>18.444734</td>\n",
       "      <td>0.596167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ERS2540950</td>\n",
       "      <td>04CN-65-106</td>\n",
       "      <td>VNII</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-106</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624139</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.274274</td>\n",
       "      <td>0.965697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ERS2540925</td>\n",
       "      <td>04CN-65-129</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-129</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99/Crypto...</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624242</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>32.821441</td>\n",
       "      <td>0.974894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ERS2540931</td>\n",
       "      <td>04CN-65-133</td>\n",
       "      <td>AD_hybrid</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>04CN-65-133</td>\n",
       "      <td>Cryptococcus neoformans var. grubii H99/Crypto...</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>positive</td>\n",
       "      <td>ERR2624415</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>44.428216</td>\n",
       "      <td>0.982900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ERS2541245</td>\n",
       "      <td>19CN-03-220-CSF</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19CN-03-220-CSF</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624247</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.227538</td>\n",
       "      <td>0.591819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ERS2541119</td>\n",
       "      <td>19CN-03-220-PAL</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19CN-03-220-PAL</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624419</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.170335</td>\n",
       "      <td>0.596044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ERS2541337</td>\n",
       "      <td>BMD_1385</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_1385</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624309</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>27.804298</td>\n",
       "      <td>0.608818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ERS2541338</td>\n",
       "      <td>BMD_1516</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_1516</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624397</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>25.317098</td>\n",
       "      <td>0.601971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ERS2541339</td>\n",
       "      <td>BMD_2014</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD_2014</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624500</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.731607</td>\n",
       "      <td>0.587846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ERS2541082</td>\n",
       "      <td>BMD1964</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD1964</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624244</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>35.464555</td>\n",
       "      <td>0.601434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ERS2541145</td>\n",
       "      <td>BMD2233</td>\n",
       "      <td>gattii</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMD2233</td>\n",
       "      <td>Cryptococcus gattii WM276</td>\n",
       "      <td>Ashton</td>\n",
       "      <td>negative</td>\n",
       "      <td>ERR2624259</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>27.246364</td>\n",
       "      <td>0.600599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample           strain    lineage    source VNI_subdivision  \\\n",
       "0   ERS2541256      04CN-03-031  AD_hybrid  Clinical             NaN   \n",
       "1   ERS2541170      04CN-03-074       VNII  Clinical             NaN   \n",
       "2   ERS2541130      04CN-03-088       VNII  Clinical             NaN   \n",
       "3   ERS2540936      04CN-63-006     gattii  Clinical             NaN   \n",
       "4   ERS2540975      04CN-63-020     gattii  Clinical             NaN   \n",
       "5   ERS2541070      04CN-63-021     gattii  Clinical             NaN   \n",
       "6   ERS2541050      04CN-64-092  AD_hybrid  Clinical             NaN   \n",
       "7   ERS2541321      04CN-65-019  AD_hybrid  Clinical             NaN   \n",
       "8   ERS2541042      04CN-65-031       VNII  Clinical             NaN   \n",
       "9   ERS2541325      04CN-65-051     gattii  Clinical             NaN   \n",
       "10  ERS2540953      04CN-65-104     gattii  Clinical             NaN   \n",
       "11  ERS2540950      04CN-65-106       VNII  Clinical             NaN   \n",
       "12  ERS2540925      04CN-65-129  AD_hybrid  Clinical             NaN   \n",
       "13  ERS2540931      04CN-65-133  AD_hybrid  Clinical             NaN   \n",
       "14  ERS2541245  19CN-03-220-CSF     gattii  Clinical             NaN   \n",
       "15  ERS2541119  19CN-03-220-PAL     gattii  Clinical             NaN   \n",
       "16  ERS2541337         BMD_1385     gattii  Clinical             NaN   \n",
       "17  ERS2541338         BMD_1516     gattii  Clinical             NaN   \n",
       "18  ERS2541339         BMD_2014     gattii  Clinical             NaN   \n",
       "19  ERS2541082          BMD1964     gattii  Clinical             NaN   \n",
       "20  ERS2541145          BMD2233     gattii  Clinical             NaN   \n",
       "\n",
       "             Lab ID                      Species ID from mash anlaysis  \\\n",
       "0       04CN-03-031  Cryptococcus neoformans var. grubii H99/Crypto...   \n",
       "1       04CN-03-074            Cryptococcus neoformans var. grubii H99   \n",
       "2       04CN-03-088            Cryptococcus neoformans var. grubii H99   \n",
       "3       04CN-63-006                          Cryptococcus gattii WM276   \n",
       "4       04CN-63-020                          Cryptococcus gattii WM276   \n",
       "5       04CN-63-021                          Cryptococcus gattii WM276   \n",
       "6       04CN-64-092  Cryptococcus neoformans var. grubii H99/Crypto...   \n",
       "7       04CN-65-019  Cryptococcus neoformans var. grubii H99/Crypto...   \n",
       "8       04CN-65-031            Cryptococcus neoformans var. grubii H99   \n",
       "9       04CN-65-051                          Cryptococcus gattii WM276   \n",
       "10      04CN-65-104                          Cryptococcus gattii WM276   \n",
       "11      04CN-65-106            Cryptococcus neoformans var. grubii H99   \n",
       "12      04CN-65-129  Cryptococcus neoformans var. grubii H99/Crypto...   \n",
       "13      04CN-65-133  Cryptococcus neoformans var. grubii H99/Crypto...   \n",
       "14  19CN-03-220-CSF                          Cryptococcus gattii WM276   \n",
       "15  19CN-03-220-PAL                          Cryptococcus gattii WM276   \n",
       "16         BMD_1385                          Cryptococcus gattii WM276   \n",
       "17         BMD_1516                          Cryptococcus gattii WM276   \n",
       "18         BMD_2014                          Cryptococcus gattii WM276   \n",
       "19          BMD1964                          Cryptococcus gattii WM276   \n",
       "20          BMD2233                          Cryptococcus gattii WM276   \n",
       "\n",
       "     Study HIV status SRA_Accession Country_of_origin Continent of Origin  \\\n",
       "0   Ashton   positive    ERR2624093           Vietnam                Asia   \n",
       "1   Ashton   positive    ERR2624467           Vietnam                Asia   \n",
       "2   Ashton   positive    ERR2624180           Vietnam                Asia   \n",
       "3   Ashton   positive    ERR2624156            Uganda              Africa   \n",
       "4   Ashton   positive    ERR2624413            Uganda              Africa   \n",
       "5   Ashton   positive    ERR2624263            Uganda              Africa   \n",
       "6   Ashton   positive    ERR2624285            Uganda              Africa   \n",
       "7   Ashton   positive    ERR2624434            Uganda              Africa   \n",
       "8   Ashton   positive    ERR2624151            Uganda              Africa   \n",
       "9   Ashton   positive    ERR2624462            Uganda              Africa   \n",
       "10  Ashton   positive    ERR2624451            Uganda              Africa   \n",
       "11  Ashton   positive    ERR2624139            Uganda              Africa   \n",
       "12  Ashton   positive    ERR2624242            Uganda              Africa   \n",
       "13  Ashton   positive    ERR2624415            Uganda              Africa   \n",
       "14  Ashton   negative    ERR2624247           Vietnam                Asia   \n",
       "15  Ashton   negative    ERR2624419           Vietnam                Asia   \n",
       "16  Ashton   negative    ERR2624309           Vietnam                Asia   \n",
       "17  Ashton   negative    ERR2624397           Vietnam                Asia   \n",
       "18  Ashton   negative    ERR2624500           Vietnam                Asia   \n",
       "19  Ashton   negative    ERR2624244           Vietnam                Asia   \n",
       "20  Ashton   negative    ERR2624259           Vietnam                Asia   \n",
       "\n",
       "    Year of Origin  Mean depth of mapping with MQ > 30 across whole genome  \\\n",
       "0           2013.0                                          45.693733        \n",
       "1           2014.0                                          43.472265        \n",
       "2           2014.0                                          45.057615        \n",
       "3           2014.0                                          30.811483        \n",
       "4           2014.0                                          31.976266        \n",
       "5           2014.0                                          30.489235        \n",
       "6           2014.0                                          32.871236        \n",
       "7           2013.0                                          78.893862        \n",
       "8           2013.0                                          31.248864        \n",
       "9           2013.0                                          44.914011        \n",
       "10          2014.0                                          18.444734        \n",
       "11             NaN                                          43.274274        \n",
       "12          2014.0                                          32.821441        \n",
       "13          2014.0                                          44.428216        \n",
       "14             NaN                                          22.227538        \n",
       "15             NaN                                          26.170335        \n",
       "16          2007.0                                          27.804298        \n",
       "17          2008.0                                          25.317098        \n",
       "18             NaN                                          20.731607        \n",
       "19          2009.0                                          35.464555        \n",
       "20          2010.0                                          27.246364        \n",
       "\n",
       "    Proportion of genome covered by at least 5 reads which mapped with MQ > 30  \n",
       "0                                            0.975470                           \n",
       "1                                            0.965882                           \n",
       "2                                            0.966880                           \n",
       "3                                            0.611245                           \n",
       "4                                            0.604924                           \n",
       "5                                            0.602266                           \n",
       "6                                            0.969852                           \n",
       "7                                            0.978027                           \n",
       "8                                            0.963734                           \n",
       "9                                            0.616338                           \n",
       "10                                           0.596167                           \n",
       "11                                           0.965697                           \n",
       "12                                           0.974894                           \n",
       "13                                           0.982900                           \n",
       "14                                           0.591819                           \n",
       "15                                           0.596044                           \n",
       "16                                           0.608818                           \n",
       "17                                           0.601971                           \n",
       "18                                           0.587846                           \n",
       "19                                           0.601434                           \n",
       "20                                           0.600599                           "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains_not_in_tree = joined[~joined['strain'].isin(tips)].reset_index(drop = True)\n",
    "strains_not_in_tree"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "widgets_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
