# Crypto_Diversity_Pipeline

Pipeline for genome assembly and analysis of data from Desjardins et al. 2017


## Broad description (new version)

Raw Illumina sequencing reads generated by Desjardins et al. 2017 were downloaded from the NIH Sequence Read Archive (BioProject ID PRJNA382844).  For each of the 387 BioSamples (corresponding strains of interest) associated with the BioProject, we created a reference based genome assembly based on aligning paired-end sequence data to the genome of the reference of the corresponding lineage (see below). In cases where there were multiple sequencing runs for a given BioSample, we concatenated the FASTQ files of the multiple runs into one.  

We used the following reference genome assemblies to align the reads of a sample of a given lineage to its corresponding reference:  
* VNI -- H99, [GCA_011801205.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_011801205.1/)  
* VNII -- VNII strain collected from cockatoo excrement , [GCA_022832995.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_022832995.1/)  
* VNBI -- Ftc146-1, [GCA_010065285.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_010065285.1/)
* VNBII -- Bt81, [GCA_023650555.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_023650555.1/)  

To create reference based genome assemblies we aligned reads to the corresponding reference genome using BWA (vXXX; Li and Durbin 2009), called variants using FreeBayes (vXXX; Garrison and Marth 2012), and generated strain-specific consensus assembly by instantiating the called variants onto the corresponding reference genome.  The read alignment, variant calling, and consensus assembly were carried out using the Snippy (https://github.com/tseemann/snippy) pipeline tool.

To have a consistent naming scheme in the annotations we "lifted over" the annotation of the strain H99 (FungiDB R65) to each of the four reference genomes using the software tool Liftoff (vXXX; Shumate and Salzberg 2020). Following construction of consensus assemblies, genome feature annotation was "lifted over" from the corresponding reference genome to each strain-specific genome.  The `-polish` option of Liftoff was employed to re-align exons in cases where the lift-over procedure resulted in start/stop codon loss or introduced an in-frame stop codon.  

Based on the polished lift-over annotation, the AGAT GTF/GFF Toolkit software (https://github.com/NBISweden/AGAT) was used to predict protein sequences for all annotated genes in each strain-specific assembly using the `agat_sp_extract_sequences.pl` script. Where multiple protein isoforms are annotated in the reference genome, we generated predictions for each isoform.



### Starting files: 
  * `files/Desjardins_Supplemental_Table_S1.csv` name and 1st line modified from from [original table](https://genome.cshlp.org/content/suppl/2017/06/05/gr.218727.116.DC1/Supplemental_Table_S1.xlsx)
  * `files/lineage_references.csv`
  * `VNI.fasta`, `VNII.fasta`, `VNBI.fasta` and `VNBII.fasta` in `references/` (accessions mentioned above)
  * `references/`[FungiDB-65_CneoformansH99_Genome.fasta](https://fungidb.org/common/downloads/release-65/CneoformansH99/fasta/data/FungiDB-65_CneoformansH99_Genome.fasta)
  * `references/`[FungiDB-65_CneoformansH99.gff](https://fungidb.org/common/downloads/release-65/CneoformansH99/gff/data/FungiDB-65_CneoformansH99.gff)
  * Table with genes to add to the chromosome plots `files/loci.csv` with:
    * Centromere delimiting-gene IDs from Janbon 2014. 
    * MAT loci protein_coding_gene IDs (everything between SXI1 and STE12):  
`awk '/SXI1/,/STE12/' references/FungiDB-65_CneoformansH99.gff.tsv | grep protein_coding_gene | cut -f13 > files/MAT.txt`.  
    * rRNA genes from the level2 primary_tag = rRNA and converting the ID into gene_ID (because the level1 primary tag is ncRNA_gene and the pattern rRNA is also in descriptions that are nor rRNA genes):  
`awk '$3 == "rRNA" {print $0}' references/FungiDB-65_CneoformansH99.gff.tsv  | cut -f13 | cut -d'-' -f1 > files/rRNA.txt`  

### Scripts used in this order:

#### Get FASTQ files of a BioProject
1. `sra-bioproject.xsh` -- given the NCBI BioProject ID `PRJNA382844`, identifies all the BioSamples associated with the project and downloads each into a folder called `srafiles/${SRSID}` where `SRSID` are SRA (Sequence Read Archive) SRS numbers. It the prefetches the srs files into uncompressed fastq files into `fastqs/`. It also creates CSV files to know which read files are paired and unpaired runs of each sample.
    * Files produced:
        
      * `files/{project_id}_biosamples.txt` -- a tab-delimited text file listing all the samples associated with the BioProject.  Three columns: SAM, Name, SRS  where SAM = BioSample Sample ID , Name = common name, SRS = SRA sequence sample identifier
        
      * `files/{project_id}_SRStoSRR.txt` -- tab-delimited text file giving mapping between SRS (Samples) IDs and SRR (Runs) IDs. A given SRS can have multiple associated SRRs. 
        
      * Each `srafiles/${SRSID}` directory will contain a number of SRR subdirectories of the form `srafiles/${SRSID}/{SRRID}` where SRA runs are "prefetched" using the SRA Tools `prefetch` command.

      * `fastqs/` directory.
    
      * `files/read_pair_table.csv` -- Columns are SRS ID, SRR ID, read pair 1, read pair 2, total size in bytes of read pair
      * `files/unpaired_fastqs.csv` -- Columns are SRS ID, SRR ID, fastq file(s)
    
#### Get proper format files for dataset
5. `get-lineage-of-samples.xsh` -- add the SRS codes to the `files/Desjardins_Supplemental_Table_S1.csv` using `read_pair_table.csv`.  
    * Files produced:
    
      * `files/sample_metadata.csv`  -- Columns are Strain, Sample, Name, Group, VNI subdivision, Mating type,Country of origin, Isolation source, Broad Project, SRA Accession, Strain description, In GWAS, Phenotyped

6. `get-removed-chromosome.sh` -- Remove mitochondrial chromosome from reference genome (`VNBI.fasta`).

7. `get-chromosome-names.sh` -- Get chromosome names from FASTA headers of each refrence.
  * Files produced:
    * `files/chromosome_names.csv`

