# Crypto_Diversity_Pipeline

Description of running the DiversityPipeline in the Desjardins et al. 2017 dataset.

## Data

The raw Illumina sequencing reads generated by Desjardins et al. 2017 were downloaded from the NIH Sequence Read Archive (BioProject ID PRJNA382844). In the cases where there were multiple sequencing runs for a given BioSample, we concatenated the FASTQ files of the multiple runs into one.  

The reference genome assemblies for each lineage are:
 * VNI: Strain H99 from [FungiDB release 65](https://fungidb.org/common/downloads/release-65/CneoformansH99/fasta/data/FungiDB-65_CneoformansH99_Genome.fasta).  
 * VNII: Strain VNII from [GCA_022832995.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_022832995.1/).  
 * VNBI: Strain Bt22, genome assembled by Marco A. Coelho.  
 * VNBII: Strain Bt81 from [GCA_023650555.1/](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_023650555.1/).

The script `get-removed-chromosome.sh` was used to remove the mitochondrial chromosome of the VNI genome.

## Config files 
* `config/metadata.csv`: Modified from from [original table](https://genome.cshlp.org/content/suppl/2017/06/05/gr.218727.116.DC1/Supplemental_Table_S1.xlsx) using the script `scripts/get-lineage-of-samples.xsh` to add the SRS codes to the `files/Desjardins_Supplemental_Table_S1.csv` using `read_pair_table.csv` (from the results of `downloading-tools`. And a new column `source` was added to simplify the information in `Isolation_source`.  
* `config/chromosomes.csv`: Table with the accession of each chromosome sequence and the common name of the chromosomes of the reference genome for each lineage. Obtained with`scripts/get-chromosome-names.sh`.  
* `config/loci.csv`: Table with genes to add to the chromosome plots:
 * Centromere delimiting-gene IDs from Janbon 2014. 
 * MAT loci protein_coding_gene IDs (everything between SXI1 and STE12):  
`awk '/SXI1/,/STE12/' references/FungiDB-65_CneoformansH99.gff.tsv | grep protein_coding_gene | cut -f13 > files/MAT.txt`.  
 * rRNA genes from the level2 primary_tag = rRNA and converting the ID into gene_ID (because the level1 primary tag is ncRNA_gene and the pattern rRNA is also in descriptions that are nor rRNA genes):  
`awk '$3 == "rRNA" {print $0}' references/FungiDB-65_CneoformansH99.gff.tsv  | cut -f13 | cut -d'-' -f1 > files/rRNA.txt`.

## Published database versions

| DiversityPipeline commit | config.yaml commit (this repo) | Shinyt App commit (ServerDatabase repo) | Date |
| :----------------------- | :----------------------------- |:--------------------------------------- |----: |
|2656cb0|57d8395|e64549a|Aug 19th 2024|
