# WeavePop_Cneoformans

Description of running the [WeavePop](https://github.com/magwenelab/WeavePop) in the Desjardins et al. 2017 and Ashton et al. 2019 datasets and the analyses of the results.

The working directory of everything described here is `WeavePop_Cneoformans/`, unless otherwise specified.  

## Desjardins

### Download and processing of data and config files

The reference genome assemblies for each lineage are:
 * VNI: Strain H99 from [FungiDB release 65](https://fungidb.org/common/downloads/release-65/CneoformansH99/).  
 * VNII: Strain VNII from [GCA_022832995.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_022832995.1/).  
 * VNBI: Strain Bt22, genome assembled by Marco A. Coelho.  
 * VNBII: Strain Bt89 from [GCA_023650575.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_023650575.1/).

The script `scripts/get-removed-chromosome.sh` was used to remove the mitochondrial chromosome of the `VNI.fasta` genome. 

The raw Illumina sequencing reads generated by Desjardins et al. 2017 were downloaded from the NIH Sequence Read Archive (BioProject ID PRJNA382844) using the workflow [download-tools](https://github.com/magwenelab/download-tools). In the cases where there were multiple sequencing runs for a given BioSample, we concatenated the FASTQ files of all the paired runs into one. This creates the FASTQ files and the table `reads_table.csv` -- Columns are `sample` (SRS ID), `run` (SRR ID), `file1` (read pair 1), `file2` (read pair 2), `file_unpaired` (unpaired reads). The unpaired reads were ignored.

The reads were cleaned with FastP using the script `scripts/fastp_clean.sh` (executed wiht `scripts/run_fastp_clean.sh`.)

The metadata was formatted with the script `scripts/prepare_desjardins_metadata.xsh`. It adds the sample ID (SRS accession) to the original metadata table and standardize the format. It uses `config/Desjardins_Supplemental_Table_S1.csv` (the name and 1st line were modified from [original table](https://genome.cshlp.org/content/suppl/2017/06/05/gr.218727.116.DC1/Supplemental_Table_S1.xlsx)) 
and `config/reads_table.csv`. It was run with the environment `misc/sra-tools.yaml`.
  * `config/metadata.csv`. 

The script `scripts/get-chromosome-names.sh` was run to obtain a table with the accession of each chromosome sequence and the common name of the chromosomes of the reference genome for each lineage:  
  * `config/chromosomes.csv`  

A table with genes of interest to add to the plots was created with the following information: The gene IDs of the centromere adjacent genes were takem from Janbon 2014. MAT loci protein_coding_gene IDs (everything between SXI1 and STE12) were taken from the VNI genome GFF. The rRNA genes (with the tags level1 ncRNA and level2 rRNA) from the VNI annotation.    
  * `config/loci.csv`

The RepBase database of consensus repetitive sequences was downloaded with the following steps to create: 
  * `config/RepBase.fasta`  
```
wget https://www.girinst.org/server/RepBase/protected/RepBase29.01.fasta.tar.gz
tar -xvzf RepBase29.01.fasta.tar.gz
cat RepBase29.01.fasta/*.ref > RepBase.fasta
cat RepBase29.01.fasta/appendix/*.ref >> RepBase.fasta
rm -rf RepBase29.01.fasta/ RepBase29.01.fasta.tar.gz 
```

### Workflow execution

```
conda activate snakemake
snakemake --profile Crypto_Desjardins/config/default
``` 

## Ashton

### Download and processing of data and config files

The script `scripts/prepare_metadata_ashton.ipynb` was used to format the metadata table. It uses the original metadata table `config/Ashton_Supplemental_Table_S1.csv` and the reads table `config/reads_table.csv`. It creates the file:
* `config/metadata.csv`

### Workflow execution

## Joint datasets Desjardins_Ashton
### Execution

## Analyses

```
analyses/
├── data # Raw data and data processed by the scripts here
├── misc
├── notebooks # Rendered versions of Quarto documents
├── results # Results of the analyses
├── scripts # Scripts, Jupyter notebooks, and Quarto documents
```
The working directory of all the Quarto documents is `analyses/`.  
They are rendered with the command: `quarto render analyses/scripts/<name>.qmd`.  
`analyses/scripts/metadata_colors.R` creates color palettes for the metadata to use in the plots.

All the input files used in the analyses come from the input or results of WeavePop in `Crypto_Desjardins`, `Crypto_Ashton` or `Crypto_Desjardins_Ashton` or from the following external data.  

External data:  
* `data/raw/CryptoDiversity_Desjardins_Tree.tre`: From [CryptoDiversity_Tree_Info](https://github.com/magwenelab/CryptoDiversity_Tree_Info/blob/main/CryptoDiversity_Desjardins_Tree.tre)
* `data/raw/2017.06.09.all_ours_and_desj.snp_sites.mod.fa.cln.tree`: Sent by Philip Ashton (Dec 5 2024).


| Analysis | Script <br /> `scripts/` | Files <br /> `data/processed/` or specified | Description |
|-----------------|-----------------|-----------------| -----------------|
| Explore Depth Profile of all Samples | `explore_depth.qmd` | `results/tables/ploidy.tsv` | Explore the depth plots to identify putative non-haploid samples to exclude from the analyses. |
| Metadata | `metadata.ipynb` | `metadata_all_H99_complete.csv`<br />  `metadata_ashton_desj_all_weavepop_final_H99.csv`<br />   `metadata_ashton_desj_vni_weavepop_final_H99.csv`  | Create new metadata tables to add the VNI subdivision information from the Ashton study to the Desjardins samples and remove the samples excluded by ploidy or quality. |
| Tree building | `merge_trees.qmd` | `tree_ashton.newick`<br />  `tree_desjardins.newick`<br />  `tree_merged.newick`<br /> Plots in `results/trees/` | Merge the trees of the Ashton and Desjardins datasets. |
| Discover aneuploidies | `aneuploidies.qmd` | `results/tables/chromosome_cnv_categories.tsv`<br /> Plots in `results/figs/`| Categorize chromosomes by coverage of CNVs |
| Plot duplications in tree |`tree_plot_cnvs.qmd`| Plots in `results/trees_dups/`| Plot the merged tree with a heatmap of duplicated chromosomes.|
|Metrics of quality and variants of final dataset | `snp_counts.qmd` | `snp_counts_desjardins.csv`<br /> `snp_counts_ashton.csv` <br /> `results/tables/per_lineage_summary_stats.tsv`| Create summary table of  mapping stats, number of variants, and CNVs. |
