# Crypto_Diversity_Pipeline

Description of running the [DiversityPipeline](https://github.com/magwenelab/DiversityPipeline) in the Desjardins et al. 2017 dataset.

## Download and preparation of data and config files

The reference genome assemblies for each lineage are:
 * VNI: Strain H99 from [FungiDB release 65](https://fungidb.org/common/downloads/release-65/CneoformansH99/).  
 * VNII: Strain VNII from [GCA_022832995.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_022832995.1/).  
 * VNBI: Strain Bt22, genome assembled by Marco A. Coelho.  
 * VNBII: Strain Bt81 from [GCA_023650555.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_023650555.1/).

The script `scripts/get-removed-chromosome.sh` was used to remove the mitochondrial chromosome of the `VNI.fasta` genome. However, the same genome without removing the mitochondrial 
chromosome was used as a main reference to annotate the rest of the genomes.  

The raw Illumina sequencing reads generated by Desjardins et al. 2017 were downloaded from the NIH Sequence Read Archive (BioProject ID PRJNA382844) using the workflow [download-tools](https://github.com/magwenelab/download-tools). In the cases where there were multiple sequencing runs for a given BioSample, we concatenated the FASTQ files of all the paired runs into one. This creates the FASTQ files and the table `files/reads_table.csv` -- Columns are `sample` (SRS ID), `run` (SRR ID), `file1` (read pair 1), `file2` (read pair 2), `file_unpaired` (unpaired reads). The unpaired reads were ignored.

The script `scripts/get-lineage-of-samples.xsh` was used to add the sample ID (SRS accession) to the original metadata table using `files/Desjardins_Supplemental_Table_S1.csv` (the name and 1st line were modified from [original table](https://genome.cshlp.org/content/suppl/2017/06/05/gr.218727.116.DC1/Supplemental_Table_S1.xlsx)) and `files/reads_table.csv` to get:  
  * `config/metadata.csv`. 

The script `scripts/get-chromosome-names.sh` was run to obtain a table with the accession of each chromosome sequence and the common name of the chromosomes of the reference genome for each lineage:  
  * `config/chromosomes.csv`  

A table with genes of interest to add to the plots was created with the following information: The gene IDs of the centromere adjacent genes were takem from Janbon 2014. MAT loci protein_coding_gene IDs (everything between SXI1 and STE12) were taken from the VNI genome GFF. The rRNA genes (with the tags level1 ncRNA and level2 rRNA) from the VNI annotation.    
  * `config/loci.csv`

The RepBase database of consensus repetitive sequences was downloaded with the following steps to create: 
  * `config/RepBase.fasta`  
```
wget https://www.girinst.org/server/RepBase/protected/RepBase29.01.fasta.tar.gz
tar -xvzf RepBase29.01.fasta.tar.gz
cat RepBase29.01.fasta/*.ref > RepBase.fasta
cat RepBase29.01.fasta/appendix/*.ref >> RepBase.fasta
rm -rf RepBase29.01.fasta/ RepBase29.01.fasta.tar.gz 
```

## Run the workflow

```
conda activate diversity
snakemake --cores 90 -p --sdm conda --keep-going`--rerun-incomplete
```

## Published database versions

History of workflow, configuration, and app commits when the database was created and published.

| Date | Workflow (DiversityPipeline repo) | Config files (this repo) | Shiny App (ServerDatabase repo) | Notes |
| :--- | :-------------------------------- | :----------------------- |:------------------------------- |-----: |
|Aug 20th 2024| 2656cb0 | 57d8395 | b8ba5c2 | The pipeline failed because of RepeatMasker (weird random error) and I had to restart the run without changing anything. |
