# WeavePop_Cneoformans

Description of running the [WeavePop](https://github.com/magwenelab/WeavePop) in the Desjardins et al. 2017 and Ashton et al. 2019 datasets and the analyses of the results.

The working directory of everything described here is `WeavePop_Cneoformans/`, unless otherwise specified.  

## Desjardins

### Download and processing of data and config files

The reference genome assemblies for each lineage are:
 * VNI: Strain H99 from [FungiDB release 65](https://fungidb.org/common/downloads/release-65/CneoformansH99/).  
 * VNII: Strain VNII from [GCA_022832995.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_022832995.1/).  
 * VNBI: Strain Bt22, genome assembled by Marco A. Coelho.  
 * VNBII: Strain Bt81 from [GCA_023650555.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_023650555.1/).

The script `scripts/get-removed-chromosome.sh` was used to remove the mitochondrial chromosome of the `VNI.fasta` genome. However, the same genome without removing the mitochondrial 
chromosome was used as a main reference to annotate the rest of the genomes.  

The raw Illumina sequencing reads generated by Desjardins et al. 2017 were downloaded from the NIH Sequence Read Archive (BioProject ID PRJNA382844) using the workflow [download-tools](https://github.com/magwenelab/download-tools). In the cases where there were multiple sequencing runs for a given BioSample, we concatenated the FASTQ files of all the paired runs into one. This creates the FASTQ files and the table `files/reads_table.csv` -- Columns are `sample` (SRS ID), `run` (SRR ID), `file1` (read pair 1), `file2` (read pair 2), `file_unpaired` (unpaired reads). The unpaired reads were ignored.

The reads were cleaned with FastP using the script `scripts/fastp_clean.sh` (executed wiht `scripts/run_fastp_clean.sh`.)

The script `scripts/get-lineage-of-samples.xsh` was used to add the sample ID (SRS accession) to the original metadata table using `files/Desjardins_Supplemental_Table_S1.csv` (the name and 1st line were modified from [original table](https://genome.cshlp.org/content/suppl/2017/06/05/gr.218727.116.DC1/Supplemental_Table_S1.xlsx)) and `files/reads_table.csv` to get:  
  * `config/metadata.csv`. 

The script `scripts/get-chromosome-names.sh` was run to obtain a table with the accession of each chromosome sequence and the common name of the chromosomes of the reference genome for each lineage:  
  * `config/chromosomes.csv`  

A table with genes of interest to add to the plots was created with the following information: The gene IDs of the centromere adjacent genes were takem from Janbon 2014. MAT loci protein_coding_gene IDs (everything between SXI1 and STE12) were taken from the VNI genome GFF. The rRNA genes (with the tags level1 ncRNA and level2 rRNA) from the VNI annotation.    
  * `config/loci.csv`

The RepBase database of consensus repetitive sequences was downloaded with the following steps to create: 
  * `config/RepBase.fasta`  
```
wget https://www.girinst.org/server/RepBase/protected/RepBase29.01.fasta.tar.gz
tar -xvzf RepBase29.01.fasta.tar.gz
cat RepBase29.01.fasta/*.ref > RepBase.fasta
cat RepBase29.01.fasta/appendix/*.ref >> RepBase.fasta
rm -rf RepBase29.01.fasta/ RepBase29.01.fasta.tar.gz 
```

### Workflow execution

```
conda activate diversity
snakemake --cores 90 -p --sdm conda --keep-going
``` 

## Ashton

### Download and processing of data and config files

### Workflow execution

## Joint datasets Desjardins_Ashton
### Execution

## Analyses

```
analyses/
├── data # Raw data and data processed by the scripts here
├── misc
├── notebooks # Rendered versions of Quarto documents
├── results # Results of the analyses
├── scripts # Scripts, Jupyter notebooks, and Quarto documents
```
The working directory of all the Quarto documents is `analyses/`.  
They are rendered with the command: `quarto render analyses/scripts/<name>.qmd`.  
`analyses/scripts/metadata_colors.R` creates color palettes for the metadata to use in the plots.

External data:  
* `data/raw/media-1.csv`: From the original Billmyre paper in [bioRxiv](https://www.biorxiv.org/content/biorxiv/early/2024/08/06/2024.07.28.605507/DC1/embed/media-1.csv?download=true).
* `data/raw/CryptoDiversity_Desjardins_Tree.tre`: From [CryptoDiversity_Tree_Info](https://github.com/magwenelab/CryptoDiversity_Tree_Info/blob/main/CryptoDiversity_Desjardins_Tree.tre)
* `data/raw/2017.06.09.all_ours_and_desj.snp_sites.mod.fa.cln.tree`: Sent by Philip Ashton (Dec 5 2024).


| Analysis | Script <br /> `scripts/` | Files <br /> `data/processed/` or specified | Description |
|-----------------|-----------------|-----------------| -----------------|
| Metadata | `metadata.ipynb` | `metadata_ashton_desj_all_weavepop_complete_info.csv`<br />  `metadata_ashton_desj_all_weavepop_H99.csv`<br />   `metadata_ashton_desj_all_weavepop_complete_info.csv`  | Create new metadata tables to add the VNI subdivision information from the Ashton study to the Desjardins samples. |
| Tree building | `merge_trees.qmd` | `tree_ashton.newick`<br />  `tree_desjardins.newick`<br />  `tree_merged.newick`<br /> Plots in `results/trees/` | Merge the trees of the Ashton and Desjardins datasets. |
|SNPs|`snps_compare_old_clean.qmd`| `snp_counts_desjardins_old.csv`<br /> `snp_counts_ashton_old.csv` | Compare number of SNPs before and after using FastP|
|  | `snp_counts.qmd` | `snp_counts_desjardins.csv`<br /> `snp_counts_ashton.csv`| Compare number of raw and filtered SNPs. |
| CNVs | `cnvs_explore.qmd` | `chromosome_lengths.tsv`| Explore relationship between CNV metrics and chromosomal depth. |
|| `cnvs_pca.qmd` || PCA of the CNV and Depth metrics. |
||`duplications_detect.qmd`||
| Ploidy | `ploidy_heterozygosity.qmd` | | Analyze ploidy with heterozygous SNPs. |
|| `ploidy_depth.qmd` <br /> ` ploidy_depth_datset.qmd` || Analyze ploidy with Mixture Models of Depth Distribution. <br /> Join the results of the previous sample-wise analysis into dataset-wide results and explore it.|
|Plot duplications in tree |`duplications_plot_tree.qmd`| `results/trees_dups/tree_merged_duplications.png`<br /> `results/trees_dups/tree_merged_duplications_12_13.png`<br /> `results/trees_dups/tree_merged_duplications_only_duplicated.png`<br /> `results/trees_dups/tree_merged_duplications_only_duplicated2.png`<br /> `results/trees_dups/tree_merged_duplications_only_duplicated3.png`| Plot the merged tree with a heatmap of duplicated chromosomes.|







## Published database versions

History of workflow, configuration, and app commits when the database was created and published.

| Date | Workflow (DiversityPipeline repo) | Config files (this repo) | Shiny App (ServerDatabase repo) | Notes |
| :--- | :-------------------------------- | :----------------------- |:------------------------------- |-----: |
|Aug 20th 2024| 2656cb0 | 57d8395 | b8ba5c2 |The pipeline failed because of RepeatMasker (weird random error) and I had to restart the run without changing anything. |
|Oct 7th 2024 | 1076166 | 07c772a | dc20b7b | This was one complete run of the analysis workflow of the Desjardins dataset, another of the Ashton dataset, and a run of the join_datasets workflow with the results of both. It is only used in the test_user. |
|Oct 16th 2024| 61e5e73 | 07c772a | 0d2035a | This was a run that updated the results of the runs of Oct 7th with the changes in the pipeline. It is being used in FungalPop |
