{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of presence of variants by impact in lineages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/FastData/czirion/ServerDatabase/scripts'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getcwd()\n",
    "os.chdir('/FastData/czirion/ServerDatabase/scripts')\n",
    "import query_database as qdb\n",
    "import pandas as pd\n",
    "import duckdb\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "mydb = '/FastData/czirion/Crypto_Desjardins/results_2024-10-22/02.Dataset/database.db'\n",
    "# mydb = '/FastData/czirion/DiversityPipeline/results_joined_R1R2/02.Dataset/database.db'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select parameters for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "impact = \"HIGH\"\n",
    "lineages = (\"VNI\", \"VNII\", \"VNBI\", \"VNBII\")\n",
    "lin_names = \"_\".join(lineages)\n",
    "phylogenetic_category = (\"monophyletic\", \"paraphyletic\", \"unique\")\n",
    "phylogenetic_category_name = \"no-poly\"\n",
    "below_percent_strains = 100\n",
    "remove_mutated_genes = True\n",
    "remove_mutated_genes_name = \"no-mut\"\n",
    "min_mapq = 20\n",
    "min_depth = 0.5\n",
    "max_depth = 1.5\n",
    "max_repeat = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Presence of variants in transcripts and strains\n",
    "Get the variants with the selected impact and the strains and transcripts where they are present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "con = duckdb.connect(database=mydb, read_only=False)\n",
    "query = f\"\"\"SELECT presence.lineage, presence.var_id, \n",
    "                samples.strain, effects.transcript_id\n",
    "            FROM presence\n",
    "            JOIN effects ON presence.var_id = effects.var_id\n",
    "            JOIN samples ON presence.sample = samples.sample \n",
    "            WHERE effects.impact = '{impact}'\n",
    "            \"\"\"\n",
    "vars_presence = con.execute(query).fetch_df()\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove variants whose effect does not affect a single transcript (e.g. fused genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>var_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_504</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00853-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_4199</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00782-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_4634</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00777-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_5130</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00768-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_5132</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00768-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168797</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_121248</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_06812-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168798</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_121305</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_06812-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168799</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_114133</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_05304-mR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168800</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_115930</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_05325-mR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168801</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_121072</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_06812-mR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168722 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       lineage            var_id strain   transcript_id\n",
       "0        VNBII     var_VNBII_504   Bt81  CNAG_00853-mR1\n",
       "1        VNBII    var_VNBII_4199   Bt81  CNAG_00782-mR1\n",
       "2        VNBII    var_VNBII_4634   Bt81  CNAG_00777-mR1\n",
       "3        VNBII    var_VNBII_5130   Bt81  CNAG_00768-mR1\n",
       "4        VNBII    var_VNBII_5132   Bt81  CNAG_00768-mR1\n",
       "...        ...               ...    ...             ...\n",
       "168797   VNBII  var_VNBII_121248   Bt39  CNAG_06812-mR1\n",
       "168798   VNBII  var_VNBII_121305   Bt39  CNAG_06812-mR1\n",
       "168799   VNBII  var_VNBII_114133   Bt39  CNAG_05304-mR2\n",
       "168800   VNBII  var_VNBII_115930   Bt39  CNAG_05325-mR1\n",
       "168801   VNBII  var_VNBII_121072   Bt39  CNAG_06812-mR1\n",
       "\n",
       "[168722 rows x 4 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vars_presence = vars_presence[vars_presence['transcript_id'].notna()]\n",
    "vars_presence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get variants in selected phylogenetic categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the phylogenetic category of each variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_phylogeny = pd.read_csv(\"/FastData/czirion/Crypto_Desjardins/fungal_pop/data/variant_status.csv\", header = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the phylogenetic category to the variants table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars_presence_phylo = vars_presence.merge(var_phylogeny, on = [\"var_id\", \"strain\"], how = \"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove variants that are not in the selected phylogenetic categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars_presence_phylo_filtered = vars_presence_phylo[vars_presence_phylo[\"category\"].isin(phylogenetic_category)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vars_presence_phylo_grouped = vars_presence_phylo_filtered.groupby(['var_id']).count()\n",
    "# vars_presence_phylo_grouped\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "# plt.figure(figsize=(10, 6))\n",
    "# plt.hist(vars_presence_phylo_grouped['strain'], bins= 100)\n",
    "# plt.xlim(0, 30)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total, filltered out, remaining variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11917, 5033, 6884)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vars_presence['var_id'].nunique(), vars_presence['var_id'].nunique() - vars_presence_phylo_filtered['var_id'].nunique(), vars_presence_phylo_filtered['var_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>var_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_504</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00853-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_4199</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00782-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_4634</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00777-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_6613</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00745-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_7802</td>\n",
       "      <td>Bt81</td>\n",
       "      <td>CNAG_00726-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168681</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_115804</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_05321-mR2</td>\n",
       "      <td>paraphyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168683</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_115974</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_05326-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168685</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_118252</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_06849-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168701</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_113216</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_05298-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168706</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>var_VNBII_115960</td>\n",
       "      <td>Bt39</td>\n",
       "      <td>CNAG_07832-mR1</td>\n",
       "      <td>monophyletic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33510 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       lineage            var_id strain   transcript_id      category\n",
       "0        VNBII     var_VNBII_504   Bt81  CNAG_00853-mR1  monophyletic\n",
       "1        VNBII    var_VNBII_4199   Bt81  CNAG_00782-mR1  monophyletic\n",
       "2        VNBII    var_VNBII_4634   Bt81  CNAG_00777-mR1  monophyletic\n",
       "6        VNBII    var_VNBII_6613   Bt81  CNAG_00745-mR1  monophyletic\n",
       "7        VNBII    var_VNBII_7802   Bt81  CNAG_00726-mR1  monophyletic\n",
       "...        ...               ...    ...             ...           ...\n",
       "168681   VNBII  var_VNBII_115804   Bt39  CNAG_05321-mR2  paraphyletic\n",
       "168683   VNBII  var_VNBII_115974   Bt39  CNAG_05326-mR1  monophyletic\n",
       "168685   VNBII  var_VNBII_118252   Bt39  CNAG_06849-mR1  monophyletic\n",
       "168701   VNBII  var_VNBII_113216   Bt39  CNAG_05298-mR1  monophyletic\n",
       "168706   VNBII  var_VNBII_115960   Bt39  CNAG_07832-mR1  monophyletic\n",
       "\n",
       "[33510 rows x 5 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vars_presence_phylo_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get variants in selected percent of strains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of strains per lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>total_strains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNI</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNII</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  lineage  total_strains\n",
       "0    VNBI            122\n",
       "1   VNBII             64\n",
       "2     VNI            185\n",
       "3    VNII             16"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con = duckdb.connect(database=mydb, read_only=False)\n",
    "query = \"SELECT lineage,strain FROM samples\"\n",
    "result = con.execute(query).fetch_df()\n",
    "con.close()\n",
    "strains_per_lineage = result.groupby(\"lineage\")['strain'].nunique().reset_index(name='total_strains')\n",
    "strains_per_lineage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number and percentage of strains with each variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars_percent = vars_presence.groupby([\"lineage\",\"var_id\", \"transcript_id\"])['strain'].nunique().reset_index(name='num_strains')\n",
    "vars_percent = vars_percent.merge(strains_per_lineage, on='lineage', how='left')\n",
    "vars_percent['percent_strains'] = (vars_percent['num_strains'] / vars_percent['total_strains']) * 100\n",
    "vars_percent['percent_strains'] = vars_percent['percent_strains'].round(2)\n",
    "vars_percent = vars_percent.drop(columns=['total_strains'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove variants present in and above the selected percentage of strains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars_percent_filtered = vars_percent[vars_percent['percent_strains'] < below_percent_strains]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total, filtered out, remaining variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11917, 170, 11747)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vars_percent['var_id'].nunique(), vars_percent['var_id'].nunique() - vars_percent_filtered['var_id'].nunique(), vars_percent_filtered['var_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>var_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_strains</th>\n",
       "      <th>percent_strains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>var_VNBI_100004</td>\n",
       "      <td>CNAG_01914-mR1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>var_VNBI_101030</td>\n",
       "      <td>CNAG_01901-mR1</td>\n",
       "      <td>31</td>\n",
       "      <td>25.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>var_VNBI_101058</td>\n",
       "      <td>CNAG_01901-mR1</td>\n",
       "      <td>52</td>\n",
       "      <td>42.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>var_VNBI_101109</td>\n",
       "      <td>CNAG_01900-mR1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>var_VNBI_101277</td>\n",
       "      <td>CNAG_01897-mR1</td>\n",
       "      <td>22</td>\n",
       "      <td>18.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13792</th>\n",
       "      <td>VNII</td>\n",
       "      <td>var_VNII_89815</td>\n",
       "      <td>CNAG_07042-mR1</td>\n",
       "      <td>11</td>\n",
       "      <td>68.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13793</th>\n",
       "      <td>VNII</td>\n",
       "      <td>var_VNII_9022</td>\n",
       "      <td>CNAG_00698-mR1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13794</th>\n",
       "      <td>VNII</td>\n",
       "      <td>var_VNII_908</td>\n",
       "      <td>CNAG_00080-mR1</td>\n",
       "      <td>12</td>\n",
       "      <td>75.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13795</th>\n",
       "      <td>VNII</td>\n",
       "      <td>var_VNII_914</td>\n",
       "      <td>CNAG_00080-mR1</td>\n",
       "      <td>12</td>\n",
       "      <td>75.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13796</th>\n",
       "      <td>VNII</td>\n",
       "      <td>var_VNII_9554</td>\n",
       "      <td>CNAG_00738-mR1</td>\n",
       "      <td>2</td>\n",
       "      <td>12.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13609 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      lineage           var_id   transcript_id  num_strains  percent_strains\n",
       "0        VNBI  var_VNBI_100004  CNAG_01914-mR1            1             0.82\n",
       "1        VNBI  var_VNBI_101030  CNAG_01901-mR1           31            25.41\n",
       "2        VNBI  var_VNBI_101058  CNAG_01901-mR1           52            42.62\n",
       "3        VNBI  var_VNBI_101109  CNAG_01900-mR1            1             0.82\n",
       "4        VNBI  var_VNBI_101277  CNAG_01897-mR1           22            18.03\n",
       "...       ...              ...             ...          ...              ...\n",
       "13792    VNII   var_VNII_89815  CNAG_07042-mR1           11            68.75\n",
       "13793    VNII    var_VNII_9022  CNAG_00698-mR1            1             6.25\n",
       "13794    VNII     var_VNII_908  CNAG_00080-mR1           12            75.00\n",
       "13795    VNII     var_VNII_914  CNAG_00080-mR1           12            75.00\n",
       "13796    VNII    var_VNII_9554  CNAG_00738-mR1            2            12.50\n",
       "\n",
       "[13609 rows x 5 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vars_percent_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Effects and annotations\n",
    "Get the table of the effects, join it with the annotation of the reference genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        SELECT samples.dataset, samples.strain, presence.sample, presence.lineage,\n",
      "            variants.var_id, chromosome_names.chromosome,\n",
      "            variants.pos AS position, variants.ref AS reference, variants.alt AS alternative,\n",
      "            effects.gene_name, effects.gene_id, effects.transcript_id,\n",
      "            effects.impact, effects.effect_type, effects.effect,\n",
      "            effects.codon_change, effects.amino_acid_change, effects.amino_acid_length,\n",
      "            effects.transcript_biotype, effects.gene_coding, effects.exon_rank,\n",
      "            mapq_depth.mean_depth_normalized, mapq_depth.mean_mapq\n",
      "        FROM variants \n",
      "        JOIN chromosome_names ON variants.accession = chromosome_names.accession\n",
      "        JOIN presence ON variants.var_id = presence.var_id\n",
      "        JOIN effects ON variants.var_id = effects.var_id\n",
      "        JOIN samples ON presence.sample = samples.sample\n",
      "        LEFT JOIN mapq_depth ON mapq_depth.feature_id = effects.transcript_id AND mapq_depth.sample = presence.sample\n",
      "        WHERE samples.dataset IN ('X',)\n",
      "        AND presence.lineage IN ('VNI', 'VNII', 'VNBI', 'VNBII')AND effects.impact IN ('HIGH',) \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "db703f6b890d46098b0ea265dae014e8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "effects = qdb.effects(db = mydb, impact = (impact,), lineage = lineages)\n",
    "keep_columns = ['lineage', 'strain', 'var_id', 'transcript_id', 'gene_id', 'impact', 'effect_type', 'mean_mapq', 'mean_depth_normalized']\n",
    "effects = effects[keep_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>strain</th>\n",
       "      <th>var_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>impact</th>\n",
       "      <th>effect_type</th>\n",
       "      <th>mean_mapq</th>\n",
       "      <th>mean_depth_normalized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>NRHc5022.ENR</td>\n",
       "      <td>var_VNBI_312896</td>\n",
       "      <td>CNAG_04268-mR1</td>\n",
       "      <td>CNAG_04268</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>SPLICE_SITE_DONOR+INTRON</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>NRHc5022.ENR</td>\n",
       "      <td>var_VNBI_312992</td>\n",
       "      <td>CNAG_04272-mR1</td>\n",
       "      <td>CNAG_04272</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>SPLICE_SITE_DONOR+INTRON</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>NRHc5022.ENR</td>\n",
       "      <td>var_VNBI_314791</td>\n",
       "      <td>CNAG_04314-mR2</td>\n",
       "      <td>CNAG_04314</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>START_LOST</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>NRHc5022.ENR</td>\n",
       "      <td>var_VNBI_315667</td>\n",
       "      <td>CNAG_04331-mR1</td>\n",
       "      <td>CNAG_04331</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>NRHc5022.ENR</td>\n",
       "      <td>var_VNBI_315876</td>\n",
       "      <td>CNAG_04337-mR1</td>\n",
       "      <td>CNAG_04337</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>SPLICE_SITE_DONOR+SPLICE_SITE_REGION+INTRON</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168797</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>PMHc1002.ENR</td>\n",
       "      <td>var_VNBII_276666</td>\n",
       "      <td>None</td>\n",
       "      <td>CNAG_04857+CNAG_04858</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>GENE_FUSION_REVERESE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168798</th>\n",
       "      <td>VNII</td>\n",
       "      <td>8-1</td>\n",
       "      <td>var_VNII_69136</td>\n",
       "      <td>None</td>\n",
       "      <td>CNAG_04739+CNAG_04740</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>GENE_FUSION_REVERESE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168799</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt142</td>\n",
       "      <td>var_VNBI_79359</td>\n",
       "      <td>None</td>\n",
       "      <td>CNAG_03878+CNAG_07482</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>GENE_FUSION_REVERESE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168800</th>\n",
       "      <td>VNI</td>\n",
       "      <td>AD6-55a</td>\n",
       "      <td>var_VNI_41401</td>\n",
       "      <td>None</td>\n",
       "      <td>CNAG_03671+CNAG_03670</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>GENE_FUSION_REVERESE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168801</th>\n",
       "      <td>VNII</td>\n",
       "      <td>AD3-9a</td>\n",
       "      <td>var_VNII_69136</td>\n",
       "      <td>None</td>\n",
       "      <td>CNAG_04739+CNAG_04740</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>GENE_FUSION_REVERESE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168802 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       lineage        strain            var_id   transcript_id  \\\n",
       "0         VNBI  NRHc5022.ENR   var_VNBI_312896  CNAG_04268-mR1   \n",
       "1         VNBI  NRHc5022.ENR   var_VNBI_312992  CNAG_04272-mR1   \n",
       "2         VNBI  NRHc5022.ENR   var_VNBI_314791  CNAG_04314-mR2   \n",
       "3         VNBI  NRHc5022.ENR   var_VNBI_315667  CNAG_04331-mR1   \n",
       "4         VNBI  NRHc5022.ENR   var_VNBI_315876  CNAG_04337-mR1   \n",
       "...        ...           ...               ...             ...   \n",
       "168797   VNBII  PMHc1002.ENR  var_VNBII_276666            None   \n",
       "168798    VNII           8-1    var_VNII_69136            None   \n",
       "168799    VNBI         Bt142    var_VNBI_79359            None   \n",
       "168800     VNI       AD6-55a     var_VNI_41401            None   \n",
       "168801    VNII        AD3-9a    var_VNII_69136            None   \n",
       "\n",
       "                      gene_id impact  \\\n",
       "0                  CNAG_04268   HIGH   \n",
       "1                  CNAG_04272   HIGH   \n",
       "2                  CNAG_04314   HIGH   \n",
       "3                  CNAG_04331   HIGH   \n",
       "4                  CNAG_04337   HIGH   \n",
       "...                       ...    ...   \n",
       "168797  CNAG_04857+CNAG_04858   HIGH   \n",
       "168798  CNAG_04739+CNAG_04740   HIGH   \n",
       "168799  CNAG_03878+CNAG_07482   HIGH   \n",
       "168800  CNAG_03671+CNAG_03670   HIGH   \n",
       "168801  CNAG_04739+CNAG_04740   HIGH   \n",
       "\n",
       "                                        effect_type  mean_mapq  \\\n",
       "0                          SPLICE_SITE_DONOR+INTRON       60.0   \n",
       "1                          SPLICE_SITE_DONOR+INTRON       60.0   \n",
       "2                                        START_LOST       60.0   \n",
       "3                                         STOP_LOST       60.0   \n",
       "4       SPLICE_SITE_DONOR+SPLICE_SITE_REGION+INTRON       60.0   \n",
       "...                                             ...        ...   \n",
       "168797                         GENE_FUSION_REVERESE        NaN   \n",
       "168798                         GENE_FUSION_REVERESE        NaN   \n",
       "168799                         GENE_FUSION_REVERESE        NaN   \n",
       "168800                         GENE_FUSION_REVERESE        NaN   \n",
       "168801                         GENE_FUSION_REVERESE        NaN   \n",
       "\n",
       "        mean_depth_normalized  \n",
       "0                        0.78  \n",
       "1                        0.86  \n",
       "2                        0.80  \n",
       "3                        0.83  \n",
       "4                        0.82  \n",
       "...                       ...  \n",
       "168797                    NaN  \n",
       "168798                    NaN  \n",
       "168799                    NaN  \n",
       "168800                    NaN  \n",
       "168801                    NaN  \n",
       "\n",
       "[168802 rows x 9 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the annotation of the reference genomes (including repeat_fraction and comparison to main reference) and join it to the strains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1c6b92071e1d412b95dd7c23948034e2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "con = duckdb.connect(database=mydb, read_only=False)\n",
    "query = f\"\"\"SELECT gff.lineage,samples.strain,\n",
    "            gff.gene_id, gff.gene_name, gff.description,\n",
    "            gff.feature_id AS transcript_id, gff.repeat_fraction,\n",
    "            gff.identical_to_main_ref, gff.start_stop_mutations\n",
    "            FROM gff\n",
    "            JOIN samples ON samples.lineage = gff.lineage\n",
    "            WHERE gff.lineage IN {lineages} and gff.primary_tag = 'mRNA'\n",
    "            \"\"\"\n",
    "annotation = con.execute(query).fetch_df()\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>strain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>repeat_fraction</th>\n",
       "      <th>identical_to_main_ref</th>\n",
       "      <th>start_stop_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06291</td>\n",
       "      <td>FPD1</td>\n",
       "      <td>polysaccharide deacetylase</td>\n",
       "      <td>CNAG_06291-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR3</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002668</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06286</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06286-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002669</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06287</td>\n",
       "      <td>GPX2</td>\n",
       "      <td>Glutathione peroxidase</td>\n",
       "      <td>CNAG_06287-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002670</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06288</td>\n",
       "      <td>None</td>\n",
       "      <td>RNA binding protein</td>\n",
       "      <td>CNAG_06288-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002671</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06289</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06289-mR1</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002672</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3002673 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        lineage                     strain     gene_id gene_name  \\\n",
       "0           VNI                     TCl-14  CNAG_06290      HXT2   \n",
       "1           VNI                     TCl-14  CNAG_06291      FPD1   \n",
       "2           VNI                     TCl-14  CNAG_06292      None   \n",
       "3           VNI                     TCl-14  CNAG_06292      None   \n",
       "4           VNI                     TCl-14  CNAG_06292      None   \n",
       "...         ...                        ...         ...       ...   \n",
       "3002668     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06286      None   \n",
       "3002669     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06287      GPX2   \n",
       "3002670     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06288      None   \n",
       "3002671     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06289      None   \n",
       "3002672     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06290      HXT2   \n",
       "\n",
       "                                   description   transcript_id  \\\n",
       "0        Low-affinity glucose transporter HXT2  CNAG_06290-mR2   \n",
       "1                   polysaccharide deacetylase  CNAG_06291-mR1   \n",
       "2                sugar transporter,  variant 2  CNAG_06292-mR1   \n",
       "3                sugar transporter,  variant 2  CNAG_06292-mR2   \n",
       "4                sugar transporter,  variant 2  CNAG_06292-mR3   \n",
       "...                                        ...             ...   \n",
       "3002668                   hypothetical protein  CNAG_06286-mR1   \n",
       "3002669                 Glutathione peroxidase  CNAG_06287-mR1   \n",
       "3002670                    RNA binding protein  CNAG_06288-mR1   \n",
       "3002671                   hypothetical protein  CNAG_06289-mR1   \n",
       "3002672  Low-affinity glucose transporter HXT2  CNAG_06290-mR1   \n",
       "\n",
       "         repeat_fraction identical_to_main_ref start_stop_mutations  \n",
       "0                   0.00                   Yes                       \n",
       "1                   0.00                   Yes                       \n",
       "2                   0.00                   Yes                       \n",
       "3                   0.00                   Yes                       \n",
       "4                   0.00                   Yes                       \n",
       "...                  ...                   ...                  ...  \n",
       "3002668             0.00                   Yes                       \n",
       "3002669             0.00                   Yes                       \n",
       "3002670             0.00                   Yes                       \n",
       "3002671             0.04                   Yes                       \n",
       "3002672             0.00                   Yes                       \n",
       "\n",
       "[3002673 rows x 9 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the annotation to the effects table to have all the posible gene-strain combinations  \n",
    "and which of them have effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation = annotation.merge(effects, on=['lineage', 'strain', 'gene_id','transcript_id'], how = \"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add status. No variant or variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation['status'] = effects_annotation['var_id'].apply(\n",
    "    lambda x: \"No variant\" if pd.isna(x) else \"Variant\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>strain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>repeat_fraction</th>\n",
       "      <th>identical_to_main_ref</th>\n",
       "      <th>start_stop_mutations</th>\n",
       "      <th>var_id</th>\n",
       "      <th>impact</th>\n",
       "      <th>effect_type</th>\n",
       "      <th>mean_mapq</th>\n",
       "      <th>mean_depth_normalized</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06291</td>\n",
       "      <td>FPD1</td>\n",
       "      <td>polysaccharide deacetylase</td>\n",
       "      <td>CNAG_06291-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR3</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040239</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06286</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06286-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040240</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06287</td>\n",
       "      <td>GPX2</td>\n",
       "      <td>Glutathione peroxidase</td>\n",
       "      <td>CNAG_06287-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040241</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06288</td>\n",
       "      <td>None</td>\n",
       "      <td>RNA binding protein</td>\n",
       "      <td>CNAG_06288-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040242</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06289</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06289-mR1</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040243</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3040244 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        lineage                     strain     gene_id gene_name  \\\n",
       "0           VNI                     TCl-14  CNAG_06290      HXT2   \n",
       "1           VNI                     TCl-14  CNAG_06291      FPD1   \n",
       "2           VNI                     TCl-14  CNAG_06292      None   \n",
       "3           VNI                     TCl-14  CNAG_06292      None   \n",
       "4           VNI                     TCl-14  CNAG_06292      None   \n",
       "...         ...                        ...         ...       ...   \n",
       "3040239     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06286      None   \n",
       "3040240     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06287      GPX2   \n",
       "3040241     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06288      None   \n",
       "3040242     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06289      None   \n",
       "3040243     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06290      HXT2   \n",
       "\n",
       "                                   description   transcript_id  \\\n",
       "0        Low-affinity glucose transporter HXT2  CNAG_06290-mR2   \n",
       "1                   polysaccharide deacetylase  CNAG_06291-mR1   \n",
       "2                sugar transporter,  variant 2  CNAG_06292-mR1   \n",
       "3                sugar transporter,  variant 2  CNAG_06292-mR2   \n",
       "4                sugar transporter,  variant 2  CNAG_06292-mR3   \n",
       "...                                        ...             ...   \n",
       "3040239                   hypothetical protein  CNAG_06286-mR1   \n",
       "3040240                 Glutathione peroxidase  CNAG_06287-mR1   \n",
       "3040241                    RNA binding protein  CNAG_06288-mR1   \n",
       "3040242                   hypothetical protein  CNAG_06289-mR1   \n",
       "3040243  Low-affinity glucose transporter HXT2  CNAG_06290-mR1   \n",
       "\n",
       "         repeat_fraction identical_to_main_ref start_stop_mutations var_id  \\\n",
       "0                   0.00                   Yes                         NaN   \n",
       "1                   0.00                   Yes                         NaN   \n",
       "2                   0.00                   Yes                         NaN   \n",
       "3                   0.00                   Yes                         NaN   \n",
       "4                   0.00                   Yes                         NaN   \n",
       "...                  ...                   ...                  ...    ...   \n",
       "3040239             0.00                   Yes                         NaN   \n",
       "3040240             0.00                   Yes                         NaN   \n",
       "3040241             0.00                   Yes                         NaN   \n",
       "3040242             0.04                   Yes                         NaN   \n",
       "3040243             0.00                   Yes                         NaN   \n",
       "\n",
       "        impact effect_type  mean_mapq  mean_depth_normalized      status  \n",
       "0          NaN         NaN        NaN                    NaN  No variant  \n",
       "1          NaN         NaN        NaN                    NaN  No variant  \n",
       "2          NaN         NaN        NaN                    NaN  No variant  \n",
       "3          NaN         NaN        NaN                    NaN  No variant  \n",
       "4          NaN         NaN        NaN                    NaN  No variant  \n",
       "...        ...         ...        ...                    ...         ...  \n",
       "3040239    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040240    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040241    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040242    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040243    NaN         NaN        NaN                    NaN  No variant  \n",
       "\n",
       "[3040244 rows x 15 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "status\n",
       "No variant    2871522\n",
       "Variant        168722\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects_annotation['status'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply filters to table of annotation with effects\n",
    "Keep only the surviving variants from the step before and the transcript quality."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by phylogenetic category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f1 = effects_annotation.copy()\n",
    "remaining_variants = effects_annotation_f1['status'] == \"Variant\"\n",
    "effects_annotation_f1 .loc[remaining_variants, 'status'] = effects_annotation_f1.loc[remaining_variants, 'var_id'].apply(\n",
    "    lambda x: \"Filtered_phylogeny\" if x not in vars_presence_phylo_filtered['var_id'].values else \"Variant\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by percentage of strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f2 = effects_annotation_f1.copy()\n",
    "remaining_variants = effects_annotation_f2['status'] == \"Variant\"\n",
    "effects_annotation_f2.loc[remaining_variants, 'status'] = effects_annotation_f2.loc[remaining_variants, 'var_id'].apply(\n",
    "    lambda x: \"Filtered_percent_strains\" if x not in vars_percent_filtered['var_id'].values else \"Variant\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by start and stop mutations in reference annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f3 = effects_annotation_f2.copy()\n",
    "remaining_variants = effects_annotation_f3['status'] == \"Variant\"\n",
    "if remove_mutated_genes:\n",
    "    effects_annotation_f3.loc[remaining_variants, 'status'] = effects_annotation_f3.apply(\n",
    "        lambda x: \"Filtered_mutation\" if x['start_stop_mutations'] != \"\" else \"Variant\",\n",
    "        axis=1)\n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by MAPQ of transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f4 = effects_annotation_f3.copy()\n",
    "remaining_variants = effects_annotation_f4['status'] == \"Variant\"\n",
    "effects_annotation_f4.loc[remaining_variants, 'status'] = effects_annotation_f4.apply(\n",
    "    lambda x: \"Filtered_mapq\" if x['mean_mapq'] < min_mapq else \"Variant\",\n",
    "    axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by depth of transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f5 = effects_annotation_f4.copy()\n",
    "remaining_variants = effects_annotation_f5['status'] == \"Variant\"\n",
    "effects_annotation_f5.loc[remaining_variants, 'status'] = effects_annotation_f5.apply(\n",
    "    lambda x: \"Filtered_depth\" if (x['mean_depth_normalized'] < min_depth) | (x['mean_depth_normalized'] > max_depth) else \"Variant\",\n",
    "    axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by fraction of repetitive sequences of the reference transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_f6 = effects_annotation_f5.copy()\n",
    "remaining_variants = effects_annotation_f6['status'] == \"Variant\"\n",
    "effects_annotation_f6.loc[remaining_variants, 'status'] = effects_annotation_f6.apply(\n",
    "    lambda x: \"Filtered_repeats\" if x['repeat_fraction'] > max_repeat else \"Variant\",\n",
    "    axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "status\n",
       "No variant                  2871522\n",
       "Filtered_phylogeny           135212\n",
       "Variant                       13271\n",
       "Filtered_percent_strains      10778\n",
       "Filtered_mutation              8994\n",
       "Filtered_repeats                259\n",
       "Filtered_depth                  196\n",
       "Filtered_mapq                    12\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects_annotation_f6['status'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>strain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>repeat_fraction</th>\n",
       "      <th>identical_to_main_ref</th>\n",
       "      <th>start_stop_mutations</th>\n",
       "      <th>var_id</th>\n",
       "      <th>impact</th>\n",
       "      <th>effect_type</th>\n",
       "      <th>mean_mapq</th>\n",
       "      <th>mean_depth_normalized</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06291</td>\n",
       "      <td>FPD1</td>\n",
       "      <td>polysaccharide deacetylase</td>\n",
       "      <td>CNAG_06291-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNI</td>\n",
       "      <td>TCl-14</td>\n",
       "      <td>CNAG_06292</td>\n",
       "      <td>None</td>\n",
       "      <td>sugar transporter,  variant 2</td>\n",
       "      <td>CNAG_06292-mR3</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040239</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06286</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06286-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040240</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06287</td>\n",
       "      <td>GPX2</td>\n",
       "      <td>Glutathione peroxidase</td>\n",
       "      <td>CNAG_06287-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040241</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06288</td>\n",
       "      <td>None</td>\n",
       "      <td>RNA binding protein</td>\n",
       "      <td>CNAG_06288-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040242</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06289</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CNAG_06289-mR1</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040243</th>\n",
       "      <td>VNI</td>\n",
       "      <td>NRHc5044.REL.INI.CLIN.ISO</td>\n",
       "      <td>CNAG_06290</td>\n",
       "      <td>HXT2</td>\n",
       "      <td>Low-affinity glucose transporter HXT2</td>\n",
       "      <td>CNAG_06290-mR1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3040244 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        lineage                     strain     gene_id gene_name  \\\n",
       "0           VNI                     TCl-14  CNAG_06290      HXT2   \n",
       "1           VNI                     TCl-14  CNAG_06291      FPD1   \n",
       "2           VNI                     TCl-14  CNAG_06292      None   \n",
       "3           VNI                     TCl-14  CNAG_06292      None   \n",
       "4           VNI                     TCl-14  CNAG_06292      None   \n",
       "...         ...                        ...         ...       ...   \n",
       "3040239     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06286      None   \n",
       "3040240     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06287      GPX2   \n",
       "3040241     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06288      None   \n",
       "3040242     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06289      None   \n",
       "3040243     VNI  NRHc5044.REL.INI.CLIN.ISO  CNAG_06290      HXT2   \n",
       "\n",
       "                                   description   transcript_id  \\\n",
       "0        Low-affinity glucose transporter HXT2  CNAG_06290-mR2   \n",
       "1                   polysaccharide deacetylase  CNAG_06291-mR1   \n",
       "2                sugar transporter,  variant 2  CNAG_06292-mR1   \n",
       "3                sugar transporter,  variant 2  CNAG_06292-mR2   \n",
       "4                sugar transporter,  variant 2  CNAG_06292-mR3   \n",
       "...                                        ...             ...   \n",
       "3040239                   hypothetical protein  CNAG_06286-mR1   \n",
       "3040240                 Glutathione peroxidase  CNAG_06287-mR1   \n",
       "3040241                    RNA binding protein  CNAG_06288-mR1   \n",
       "3040242                   hypothetical protein  CNAG_06289-mR1   \n",
       "3040243  Low-affinity glucose transporter HXT2  CNAG_06290-mR1   \n",
       "\n",
       "         repeat_fraction identical_to_main_ref start_stop_mutations var_id  \\\n",
       "0                   0.00                   Yes                         NaN   \n",
       "1                   0.00                   Yes                         NaN   \n",
       "2                   0.00                   Yes                         NaN   \n",
       "3                   0.00                   Yes                         NaN   \n",
       "4                   0.00                   Yes                         NaN   \n",
       "...                  ...                   ...                  ...    ...   \n",
       "3040239             0.00                   Yes                         NaN   \n",
       "3040240             0.00                   Yes                         NaN   \n",
       "3040241             0.00                   Yes                         NaN   \n",
       "3040242             0.04                   Yes                         NaN   \n",
       "3040243             0.00                   Yes                         NaN   \n",
       "\n",
       "        impact effect_type  mean_mapq  mean_depth_normalized      status  \n",
       "0          NaN         NaN        NaN                    NaN  No variant  \n",
       "1          NaN         NaN        NaN                    NaN  No variant  \n",
       "2          NaN         NaN        NaN                    NaN  No variant  \n",
       "3          NaN         NaN        NaN                    NaN  No variant  \n",
       "4          NaN         NaN        NaN                    NaN  No variant  \n",
       "...        ...         ...        ...                    ...         ...  \n",
       "3040239    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040240    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040241    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040242    NaN         NaN        NaN                    NaN  No variant  \n",
       "3040243    NaN         NaN        NaN                    NaN  No variant  \n",
       "\n",
       "[3040244 rows x 15 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects_annotation_f6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count the variants per gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove the variant IDs of the Filtered rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_variants = effects_annotation_f6.copy()\n",
    "effects_annotation_variants['var_id_surviving'] = effects_annotation_variants['var_id']\n",
    "effects_annotation_variants.loc[effects_annotation_variants['status'] != 'Variant', 'var_id_surviving'] = float('nan')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total variants, filtered out, remaining variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11917, 7015, 4902)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effects_annotation_variants['var_id'].nunique(), effects_annotation_variants['var_id'].nunique() -  effects_annotation_variants['var_id_surviving'].nunique(),  effects_annotation_variants['var_id_surviving'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Group by gene and strain and count the number of variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "effects_annotation_variants['gene_name'] = effects_annotation_variants['gene_name'].apply(lambda x: \"\" if pd.isna(x) else x)\n",
    "gene_strain_num_vars = effects_annotation_variants.groupby([\"lineage\", \"strain\", \"gene_id\", \"gene_name\", \"description\"])['var_id_surviving'].nunique().reset_index(name='num_vars')\n",
    "gene_strain_num_vars['presence_vars'] = gene_strain_num_vars['num_vars'].apply(lambda x: 0 if x == 0 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>strain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>presence_vars</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt102</td>\n",
       "      <td>CNAG_00001</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt102</td>\n",
       "      <td>CNAG_00006</td>\n",
       "      <td></td>\n",
       "      <td>prohibitin-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt102</td>\n",
       "      <td>CNAG_00007</td>\n",
       "      <td></td>\n",
       "      <td>methionine-tRNA ligase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt102</td>\n",
       "      <td>CNAG_00008</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>Bt102</td>\n",
       "      <td>CNAG_00009</td>\n",
       "      <td></td>\n",
       "      <td>calcium-binding protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677521</th>\n",
       "      <td>VNII</td>\n",
       "      <td>WM626</td>\n",
       "      <td>CNAG_08025</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677522</th>\n",
       "      <td>VNII</td>\n",
       "      <td>WM626</td>\n",
       "      <td>CNAG_08026</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677523</th>\n",
       "      <td>VNII</td>\n",
       "      <td>WM626</td>\n",
       "      <td>CNAG_08027</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677524</th>\n",
       "      <td>VNII</td>\n",
       "      <td>WM626</td>\n",
       "      <td>CNAG_08028</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677525</th>\n",
       "      <td>VNII</td>\n",
       "      <td>WM626</td>\n",
       "      <td>CNAG_08029</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2677526 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        lineage strain     gene_id gene_name              description  \\\n",
       "0          VNBI  Bt102  CNAG_00001               hypothetical protein   \n",
       "1          VNBI  Bt102  CNAG_00006                       prohibitin-2   \n",
       "2          VNBI  Bt102  CNAG_00007             methionine-tRNA ligase   \n",
       "3          VNBI  Bt102  CNAG_00008               hypothetical protein   \n",
       "4          VNBI  Bt102  CNAG_00009            calcium-binding protein   \n",
       "...         ...    ...         ...       ...                      ...   \n",
       "2677521    VNII  WM626  CNAG_08025               hypothetical protein   \n",
       "2677522    VNII  WM626  CNAG_08026               hypothetical protein   \n",
       "2677523    VNII  WM626  CNAG_08027               hypothetical protein   \n",
       "2677524    VNII  WM626  CNAG_08028               hypothetical protein   \n",
       "2677525    VNII  WM626  CNAG_08029               hypothetical protein   \n",
       "\n",
       "         num_vars  presence_vars  \n",
       "0               0              0  \n",
       "1               0              0  \n",
       "2               0              0  \n",
       "3               0              0  \n",
       "4               0              0  \n",
       "...           ...            ...  \n",
       "2677521         0              0  \n",
       "2677522         0              0  \n",
       "2677523         0              0  \n",
       "2677524         0              0  \n",
       "2677525         0              0  \n",
       "\n",
       "[2677526 rows x 7 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_strain_num_vars"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write the table of variants per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars_per_gene = gene_strain_num_vars.drop(['gene_name', 'description'], axis=1)\n",
    "vars_per_gene.to_csv(f'/FastData/czirion/Crypto_Desjardins/fungal_pop/data/variants_per_gene/{impact}_{lin_names}_{phylogenetic_category_name}_{below_percent_strains}_{remove_mutated_genes_name}_{min_mapq}_{min_depth}_{max_depth}_{max_repeat}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count the number of strains with variants in each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_strains_per_gene = gene_strain_num_vars.groupby([\"lineage\", \"gene_id\", 'gene_name', 'description'])['presence_vars'].sum().reset_index(name='num_strains')\n",
    "num_strains_per_gene.sort_values(by='num_strains', ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_strains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6456</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>CNAG_07553</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>polycomb protein e(z)</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1754</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>CNAG_01973</td>\n",
       "      <td>ZFC2</td>\n",
       "      <td>C2H2 zinc finger protein Zas1A</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3442</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>CNAG_03876</td>\n",
       "      <td></td>\n",
       "      <td>Ras family protein</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5966</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>CNAG_06695</td>\n",
       "      <td>MAR1</td>\n",
       "      <td>Macrophage activating regulator of cell wall-1</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5744</th>\n",
       "      <td>VNBI</td>\n",
       "      <td>CNAG_06438</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9903</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>CNAG_03405</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9902</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>CNAG_03404</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein,  variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9901</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>CNAG_03403</td>\n",
       "      <td></td>\n",
       "      <td>mitochondrial protein</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9900</th>\n",
       "      <td>VNBII</td>\n",
       "      <td>CNAG_03402</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27557</th>\n",
       "      <td>VNII</td>\n",
       "      <td>CNAG_08029</td>\n",
       "      <td></td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27558 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      lineage     gene_id gene_name  \\\n",
       "6456     VNBI  CNAG_07553      EZH2   \n",
       "1754     VNBI  CNAG_01973      ZFC2   \n",
       "3442     VNBI  CNAG_03876             \n",
       "5966     VNBI  CNAG_06695      MAR1   \n",
       "5744     VNBI  CNAG_06438             \n",
       "...       ...         ...       ...   \n",
       "9903    VNBII  CNAG_03405             \n",
       "9902    VNBII  CNAG_03404             \n",
       "9901    VNBII  CNAG_03403             \n",
       "9900    VNBII  CNAG_03402             \n",
       "27557    VNII  CNAG_08029             \n",
       "\n",
       "                                          description  num_strains  \n",
       "6456                            polycomb protein e(z)          120  \n",
       "1754                   C2H2 zinc finger protein Zas1A          120  \n",
       "3442                               Ras family protein          120  \n",
       "5966   Macrophage activating regulator of cell wall-1          119  \n",
       "5744                             hypothetical protein          119  \n",
       "...                                               ...          ...  \n",
       "9903                             hypothetical protein            0  \n",
       "9902                   hypothetical protein,  variant            0  \n",
       "9901                            mitochondrial protein            0  \n",
       "9900                             hypothetical protein            0  \n",
       "27557                            hypothetical protein            0  \n",
       "\n",
       "[27558 rows x 5 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_strains_per_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the presence of variants per strain to pass to the phylogenetic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# con = duckdb.connect(database=mydb, read_only=False)\n",
    "# query = f\"\"\"SELECT presence.var_id, samples.strain\n",
    "#             FROM presence\n",
    "#             JOIN strains ON presence.sample = samples.sample \n",
    "#             \"\"\"\n",
    "# vars_presence = con.execute(query).fetch_df()\n",
    "# con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vars_presence.to_csv(f'/FastData/czirion/Crypto_Desjardins/fungal_pop/data/vars_presence.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "widgets_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
