---
title: "Identification of Variants in the cAMP pathway"
author: "Claudia Zirión-Martínez"
date: "2025-05-23"

fig-height: 7
fig-width: 9

format: 
  html:
    page-layout: full
---

## Query cAMP variants and annotation

Use WeavePop-CLI to get the annotation of the mRNA features of the cAMP genes and the variants of all samples in those genes.

* PDE1  
* PKA1  
* PKA2  
* CAC1  
* PKR1  

```{bash, eval=FALSE}
#| code-fold: false
conda activate shiny
python query_database/weavepop_cli.py annotation \
    --db Crypto_Desjardins_Ashton/results/02.Dataset/database.db \
    --gene_name PDE1,PKA1,PKA2,CAC1,PKR1 \
    --feature_type mRNA \
    > analyses/data/processed/variants/camp_annotation.tsv

python query_database/weavepop_cli.py variants \
    --db Crypto_Desjardins_Ashton/results/02.Dataset/database.db \
    --gene_name PDE1,PKA1,PKA2,CAC1,PKR1 \
    > analyses/data/processed/variants/camp_variants.tsv
```

## Setup {#setup}

```{r, eval=FALSE, echo=FALSE}
setwd("/FastData/czirion/WeavePop_Cneoformans/analyses")
```

Libraries

```{r libraries, output=FALSE}
library(tidyverse)
library(patchwork)
library(ggnewscale)
library(RColorBrewer)
library(svglite)
```

Paths

Use the metadata table generated by `analyses/scripts/metadata.ipynb`  
and the `cnv_chromosomes.tsv` of the joined dataset generated by WeavePop.
```{r paths}
metadata_path <-
    "data/processed/metadata_ashton_desj_all_weavepop_H99.csv"
cnv_chroms_path <-
    "../Crypto_Desjardins_Ashton/results/02.Dataset/cnv/cnv_chromosomes.tsv"
annotation_path <- "data/processed/variants/camp_annotation.tsv"
variants_path <- "data/processed/variants/camp_variants.tsv"
```
## Prepare dataset {#prepare-dataset}

### Metadata {#metadata}

Use the metadata table that has all the samples included in the final **Crypto_Desjardins_Ashton** dataset (n = 1055) and H99.  
Select needed columns, remove H99 and get the number of samples per dataset and lineage, per lineage, per lineage and source and total.

```{r, message=FALSE, warning=FALSE}
metadata <- read.delim(
    metadata_path,
    header=TRUE,
    sep=",",
    stringsAsFactor = TRUE)
metadata <- metadata %>%
    select(sample, strain, source, lineage, dataset, vni_subdivision)%>%
    filter(!strain == "H99") %>%
    group_by(dataset, lineage)%>%
    mutate(samples_in_dataset_lineage = n_distinct(sample))%>%
    ungroup() %>%
    group_by(lineage)%>%
    mutate(samples_in_lineage = n_distinct(sample))%>%
    ungroup() %>%
    group_by(vni_subdivision)%>%
    mutate(samples_in_sublineage = n_distinct(sample))%>%
    ungroup() %>%
    group_by(lineage,source,samples_in_lineage) %>%
    mutate(samples_in_lineage_source = n_distinct(sample))%>%
    ungroup()%>%
    mutate(total_samples = n_distinct(sample))%>%
    droplevels()
```

### Variants
```{r}
variants <- read.delim(
    variants_path,
    header=TRUE,
    sep="\t",
    stringsAsFactor = TRUE)

variants$impact <- factor(variants$impact, levels = c("HIGH", "MODERATE", "LOW", "MODIFIER"))
variants$lineage <- factor(variants$lineage, levels = c("VNI", "VNII", "VNBI", "VNBII"))

variants <- left_join(variants, metadata, by = c("sample", "strain", "lineage"))
```

### Annotation
```{r}
annotation <- read.delim(
    annotation_path,
    header=TRUE,
    sep="\t",
    stringsAsFactor = TRUE)
annotation$lineage <- factor(annotation$lineage, levels = c("VNI", "VNII", "VNBI", "VNBII"))
```


## Explore the annotation

Number of isoforms per lineage per gene

```{r, message = FALSE, warning = FALSE}
annotation %>% 
    group_by(lineage, gene_name)%>%
    summarize(n_isoforms = n_distinct(feature_id))%>%
    pivot_wider(names_from = gene_name, values_from = n_isoforms)
```

Isoforms identical to VNI
```{r, message = FALSE, warning = FALSE}
annotation %>% 
    select(lineage, gene_name, feature_id, identical_to_main_ref, start_stop_mutations)%>%
    arrange(lineage, gene_name)
```
The VNII reference has a `missing_start_codon` in CAC1 and a `missing_stop_codon` in PDE1 mR2.

## Explore the number of variants per gene and number of samples with variants per gene

```{r, message = FALSE, warning = FALSE}
gene_lin_impact <- variants %>%
    group_by(lineage, impact, gene_name, samples_in_lineage)%>%
    arrange(gene_name,impact,lineage)%>%
    summarize(n_samples = n_distinct(sample),
                n_vars = n_distinct(var_id))%>%
    mutate(percent_samples = (n_samples / samples_in_lineage) *100)
```

```{r}
ggplot(gene_lin_impact)+
    geom_tile(aes(x = gene_name, y = lineage, fill = percent_samples))+
    geom_text(aes(x = gene_name, y = lineage, label = n_samples))+
    facet_wrap(~impact)+
    scale_fill_viridis_c(direction = -1)+
    theme_minimal()
```
```{r}
ggplot(gene_lin_impact)+
    geom_col(aes(x = gene_name, y = percent_samples, fill = lineage), position = "dodge")+
    facet_wrap(~impact)+
    theme_minimal()
```

## Explore the number of samples with each variant

```{r, message = FALSE, warning = FALSE}
vars_gene_lin_impact <- variants %>%
    group_by(lineage, impact, gene_name, var_id, samples_in_lineage)%>%
    arrange(gene_name,impact,lineage)%>%
    summarize(n_samples = n_distinct(sample))%>%
    mutate(percent_samples = round((n_samples / samples_in_lineage) *100, 1))
```

```{r, message = FALSE, warning = FALSE}
vars_gene_lin_impact_percent <- vars_gene_lin_impact %>%
    group_by(lineage, gene_name, impact, percent_samples)%>%
    summarize(n_vars = n_distinct(var_id))%>%
    arrange(desc(percent_samples))
```